File: packages/kit/src/external.ts
export interface VirtualTableSpec {
	name: string;
	module: string;
	args?: string[];
}

export function virtualTable(
	name: string,
	module: string,
	args: string[] = []
): VirtualTableSpec {
	return { name, module, args };
}

export function quoteIdent(name: string): string {
	return `"${name.replaceAll('"', '""')}"`;
}

export function createVirtualTableSql(spec: VirtualTableSpec): string {
	const args = spec.args?.join(', ') ?? '';
	const suffix = args.length > 0 ? `(${args})` : '';
	return `CREATE VIRTUAL TABLE ${quoteIdent(spec.name)} USING ${quoteIdent(spec.module)}${suffix}`;
}

export function dropVirtualTableSql(name: string): string {
	return `DROP TABLE ${quoteIdent(name)}`;
}

File: crates/mongreldb-kit-core/src/external.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VirtualTableSpec {
    pub name: String,
    pub module: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args: Vec<String>,
}

impl VirtualTableSpec {
    pub fn new(
        name: impl Into<String>,
        module: impl Into<String>,
        args: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            name: name.into(),
            module: module.into(),
            args: args.into_iter().map(Into::into).collect(),
        }
    }

    pub fn create_sql(&self) -> String {
        let args = self.args.join(", ");
        if args.is_empty() {
            format!(
                "CREATE VIRTUAL TABLE {} USING {}",
                quote_ident(&self.name),
                quote_ident(&self.module)
            )
        } else {
            format!(
                "CREATE VIRTUAL TABLE {} USING {}({args})",
                quote_ident(&self.name),
                quote_ident(&self.module)
            )
        }
    }

    pub fn drop_sql(&self) -> String {
        format!("DROP TABLE {}", quote_ident(&self.name))
    }
}

pub fn quote_ident(name: &str) -> String {
    format!("\"{}\"", name.replace('"', "\"\""))
}
